### ISSUE TITLE
ActiveRecord::Dirty does not work well with Transactions leaving model in an inconsistent state

### ISSUE THREAD

**Issue by prikha** [rails/rails#34637]
### Steps to reproduce

To be short: Changes marked as applied but are not committed. So rollback leaves a model in an inconsistent state.

The real issues arise when you have `partial_writes` enabled and have a database error while committing your transaction.

```rb
    user = User.new
    user.email = "unique@email.com"
    user.full_name = "Another New Topic"
    changes_before_rollback = user.changes

    User.transaction do
      user.save!
      raise ActiveRecord::Rollback
    end

    user.changes #=> {} but should be a hash with email and full_name changes
    user.changes_before_rollback == user.changes # => false but should logically be true 
```

If one would decide to retry the transaction. It would lead to the following place.

https://github.com/rails/rails/blob/a72ac3cfec9956ccf093d870705d3dc22ced090e/activerecord/lib/active_record/attribute_methods/dirty.rb#L179-L181

We have `partial_writes` enabled by default. So it would pick `keys_for_partial_write` which are just keys from `changes`. It would be blank on a retry attempt.

Why does it happen? I think the root is here: we mark changes as applied not caring about the transaction state.

https://github.com/rails/rails/blob/a72ac3cfec9956ccf093d870705d3dc22ced090e/activerecord/lib/active_record/attribute_methods/dirty.rb#L173-L177

There is a related issue: https://github.com/rails/rails/issues/33443 having the same root cause I believe.

### Expected behavior

We have the following examples in our WIP, but moving bits around breaks a lot of specs including associations:

```rb
# activerecord/test/cases/transactions_test.rb

  def test_rollback_of_changed_attributes
    topic = Topic.new
    topic.title = "New Topic"
    changes_before_rollback = topic.changes
     Topic.transaction do
      topic.save!
      raise ActiveRecord::Rollback
    end
     changes = topic.changes.to_a
     changes_before_rollback.to_a.each do |value|
      assert_includes(changes, value)
    end
  end

  def test_rollback_of_changed_attributes_for_persisted_record
    topic = Topic.create
    topic.title = "New Topic"
     Topic.transaction do
      topic.save!
      raise ActiveRecord::Rollback
    end
     topic.save!
     assert_equal(topic.title, "New Topic")
  end
```

I would be super happy to involve anyone who is familiar with AR code and explicitly with `ActiveRecord::Dirty` + `Transactions`. We have attempted to fix this on our own but didn't reach any success. If @rafaelfranca @sgrif @amatsuda @kamipo you guys have any ideas on the right way of fixing this I would greatly appreciate!

### Actual behavior

Now on a rollback, we loose track of the changes in the model. That leads to random errors(e.g. not null column violation) in case of retry.

### System configuration
**Rails version**: master

**Ruby version**: 2.5.3

**Comment by gmcgibbon** [comment]
Duplicate of https://github.com/rails/rails/issues/15018. I'm trying to fix this in https://github.com/rails/rails/pull/34504.

**Comment by pirj** [comment]
Apparently fixed in #35987
